# IIC2433 - Proyecto semestral


Este notebook se creó usando Google Colab, en esta plataforma es necesario correr la siguiente celda para forzar la actualización de la librería plotly. De otra forma, no se podrán ejecutar correctamente las celdas correspondientes a la Visualización en Mapas

In [ ]:
pip install -U plotly

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
from functools import reduce
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

## 1. Definición del problema y su impacto


Queremos responder a la pregunta: "Qué comunas se comportan de manera similar en su respuesta al COVID-19?". Para esto usaremos datos que entregan información respecto a la evolución y respuesta de COVID-19 en cada comuna de Chile. Estos datos provienen del [repositorio del Ministerio de Ciencia](https://github.com/MinCiencia/Datos-COVID19). 

Encontrar la respuesta a este problema puede ser de utilidad para las autoridades del Ministerio de Salud y el Ministerio de Ciencia, que utilizan la evolución de casos por comuna y la tasa de incidencia para decretar si una comuna debe presentar modificaciones en el plan Paso a Paso. 

Nuestra hipótesis es que se pueden formar grupos de comunas que han tenido un comportamiento similar en sus cifras de COVID-19. Además de los casos totales y la tasas de incidencia se podrían considerar otros factores, como la cobertura del testeo y la cantidad de tests por búsqueda activa de casos.

Estos nuevos grupos de comportamiento similar pueden ser utilizados como las unidades territoriales a las que se aplican las medidas sanitarias del plan Paso a Paso, permitiendo así que cada comuna tenga las medidas sanitarias más adecuadas a su comportamiento histórico.

El análisis lo vamos a realizar con todas las comunas del país, pero también queremos analizar solo las comunas de la región metropolitana.

## 2. Datos

Descargamos los datos de COVID-19, están disponibles en un gist de github, pero realmente vienen de un [repositorio del Ministerio de Ciencia](https://github.com/MinCiencia/Datos-COVID19). A continuación se presentan los links a las descripciones de los datos a usar: 

- [Casos totales por comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto1)
- [Movilidad en chile por comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto33)
- [Fallecidos por comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto38)
- [índice de positividad test RT-PRC por comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto65)
- [Cobertura de testeo por comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto66)
- [Cantidad de Test por búsqueda activa de casos (BAC) por Comuna](https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto64)

Los archivos csv que serán utilizados se pueden encontrar en las url de las siguientes celdas:


In [ ]:
url_total_cases = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto1/Covid-19.csv'
url_mobility = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto33/IndiceDeMovilidad-IM.csv'
url_deceased = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto38/CasosFallecidosPorComuna.csv'
url_PCR_positivity = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto65/PositividadPorComuna.csv'
url_test_coverage = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto66/CoberturaPorComuna.csv'
url_BAC = 'https://raw.githubusercontent.com/MinCiencia/Datos-COVID19/master/output/producto64/BACPorComuna.csv'

## 3. Preprocesamiento

Definimos los dataframes

In [ ]:
df_total_cases = pd.read_csv(url_total_cases)
df_mobility = pd.read_csv(url_mobility)
df_deceased = pd.read_csv(url_deceased)
df_PCR_positivity = pd.read_csv(url_PCR_positivity)
df_test_coverage = pd.read_csv(url_test_coverage)
df_BAC = pd.read_csv(url_BAC)

Realizamos una función para agrupar columas de acuerdo a la fecha

In [ ]:
def create_regular_expressions():
  years = ['2020', '2021']
  months = ['01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12']
  regex_expressions = []
  for year in years:
    for month in months:
      regex_expressions.append(year + '-' + month + '-.*')
  return regex_expressions

def group_date_columns(df, column_name):
  reg_expressions = create_regular_expressions()
  # new_df = pd.DataFrame()
  #new_df['comuna'] = df['Comuna']
  new_df = df[['Region', 'Comuna', 'Codigo comuna', 'Poblacion']]
  
  for exp in reg_expressions:
    # Creamos las columnas nuevas con el promedio del mes
    new_df[column_name + '_' + exp[:7]] = df.filter(regex=exp).mean(axis=1)
  
  # Eliminamos las fechas en que no tenemos datos
  new_df.drop(columns=[f"{column_name}_2020-01", f"{column_name}_2020-02", f"{column_name}_2021-11", f"{column_name}_2021-12"], inplace=True)
  # Los demás promedios los llenamos con ceros
  new_df.fillna(0, inplace=True)
  return new_df

Agrupamos los datos de fechas de cada df


In [ ]:
df_total_cases_info = group_date_columns(df_total_cases, 'total_cases')
df_mobility_info = group_date_columns(df_mobility, 'mobility')
df_deceased_info = group_date_columns(df_deceased, 'deceased')
df_PCR_positivity_info = group_date_columns(df_PCR_positivity, 'PRC_positivity')
df_test_coverage_info = group_date_columns(df_test_coverage, 'test_coverage')
df_BAC_info = group_date_columns(df_BAC, 'BAC')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4174: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,
/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4327: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  downcast=downcast,


Hacemos merge de los df por comuna para obtener el df con toda la información


In [ ]:
dfs = [df_total_cases_info, df_mobility_info, df_deceased_info, df_PCR_positivity_info, df_test_coverage_info, df_BAC_info]
df_merged = reduce(lambda left,right: pd.merge(left, right, on=['Codigo comuna'], how='inner'), dfs)
df_merged

Region_x          Comuna_x  ...  BAC_2021-09  BAC_2021-10
0    Arica y Parinacota             Arica  ...          0.0          0.0
1    Arica y Parinacota         Camarones  ...          0.0          0.0
2    Arica y Parinacota     General Lagos  ...          0.0          0.0
3    Arica y Parinacota             Putre  ...          0.0          0.0
4              Tarapacá     Alto Hospicio  ...          0.0          0.0
..                  ...               ...  ...          ...          ...
339          Magallanes      Punta Arenas  ...          0.0          0.0
340          Magallanes         Rio Verde  ...          0.0          0.0
341          Magallanes      San Gregorio  ...          0.0          0.0
342          Magallanes          Timaukel  ...          0.0          0.0
343          Magallanes  Torres del Paine  ...          0.0          0.0

[344 rows x 139 columns]

Creamos un df que será el que contiene toda la información y eliminamos las columnas duplicadas de región.

In [ ]:
df_master = df_merged
df_master.drop(columns=['Region_y'], inplace=True)
df_master.drop(columns=['Comuna_y'], inplace=True)
dedup_names = pd.io.parsers.ParserBase({'names':df_master.columns})._maybe_dedup_names(df_master.columns)
df_master.set_axis(dedup_names, axis=1, inplace=True)
df_master.drop(columns=['Region_x.1', 'Region_x.2', 'Comuna_x.1', 'Comuna_x.2'], inplace=True)
df_master.rename(columns={'Region_x': 'Region', 'Comuna_x': 'Comuna'}, inplace=True)

In [ ]:
# Vemos que ya no hay comunas que empiecen con "Desconocido"
df_merged[df_merged['Comuna'].str.startswith('Desconocido')]

Empty DataFrame
Columns: [Region, Comuna, Codigo comuna, Poblacion_x, total_cases_2020-03, total_cases_2020-04, total_cases_2020-05, total_cases_2020-06, total_cases_2020-07, total_cases_2020-08, total_cases_2020-09, total_cases_2020-10, total_cases_2020-11, total_cases_2020-12, total_cases_2021-01, total_cases_2021-02, total_cases_2021-03, total_cases_2021-04, total_cases_2021-05, total_cases_2021-06, total_cases_2021-07, total_cases_2021-08, total_cases_2021-09, total_cases_2021-10, Poblacion_y, mobility_2020-03, mobility_2020-04, mobility_2020-05, mobility_2020-06, mobility_2020-07, mobility_2020-08, mobility_2020-09, mobility_2020-10, mobility_2020-11, mobility_2020-12, mobility_2021-01, mobility_2021-02, mobility_2021-03, mobility_2021-04, mobility_2021-05, mobility_2021-06, mobility_2021-07, mobility_2021-08, mobility_2021-09, mobility_2021-10, Poblacion_x.1, deceased_2020-03, deceased_2020-04, deceased_2020-05, deceased_2020-06, deceased_2020-07, deceased_2020-08, deceased_2020-09, deceased_2020-10, deceased_2020-11, deceased_2020-12, deceased_2021-01, deceased_2021-02, deceased_2021-03, deceased_2021-04, deceased_2021-05, deceased_2021-06, deceased_2021-07, deceased_2021-08, deceased_2021-09, deceased_2021-10, Poblacion_y.1, PRC_positivity_2020-03, PRC_positivity_2020-04, PRC_positivity_2020-05, PRC_positivity_2020-06, PRC_positivity_2020-07, PRC_positivity_2020-08, PRC_positivity_2020-09, PRC_positivity_2020-10, PRC_positivity_2020-11, PRC_positivity_2020-12, PRC_positivity_2021-01, PRC_positivity_2021-02, PRC_positivity_2021-03, PRC_positivity_2021-04, PRC_positivity_2021-05, PRC_positivity_2021-06, PRC_positivity_2021-07, PRC_positivity_2021-08, PRC_positivity_2021-09, PRC_positivity_2021-10, Poblacion_x.2, test_coverage_2020-03, test_coverage_2020-04, test_coverage_2020-05, test_coverage_2020-06, test_coverage_2020-07, test_coverage_2020-08, test_coverage_2020-09, test_coverage_2020-10, test_coverage_2020-11, test_coverage_2020-12, test_coverage_2021-01, test_coverage_2021-02, ...]
Index: []

[0 rows x 129 columns]

In [ ]:
# No hay nulos
rows_with_NaN = df_merged[df_merged.isnull().any(axis=1)]
rows_with_NaN

Empty DataFrame
Columns: [Region, Comuna, Codigo comuna, Poblacion_x, total_cases_2020-03, total_cases_2020-04, total_cases_2020-05, total_cases_2020-06, total_cases_2020-07, total_cases_2020-08, total_cases_2020-09, total_cases_2020-10, total_cases_2020-11, total_cases_2020-12, total_cases_2021-01, total_cases_2021-02, total_cases_2021-03, total_cases_2021-04, total_cases_2021-05, total_cases_2021-06, total_cases_2021-07, total_cases_2021-08, total_cases_2021-09, total_cases_2021-10, Poblacion_y, mobility_2020-03, mobility_2020-04, mobility_2020-05, mobility_2020-06, mobility_2020-07, mobility_2020-08, mobility_2020-09, mobility_2020-10, mobility_2020-11, mobility_2020-12, mobility_2021-01, mobility_2021-02, mobility_2021-03, mobility_2021-04, mobility_2021-05, mobility_2021-06, mobility_2021-07, mobility_2021-08, mobility_2021-09, mobility_2021-10, Poblacion_x.1, deceased_2020-03, deceased_2020-04, deceased_2020-05, deceased_2020-06, deceased_2020-07, deceased_2020-08, deceased_2020-09, deceased_2020-10, deceased_2020-11, deceased_2020-12, deceased_2021-01, deceased_2021-02, deceased_2021-03, deceased_2021-04, deceased_2021-05, deceased_2021-06, deceased_2021-07, deceased_2021-08, deceased_2021-09, deceased_2021-10, Poblacion_y.1, PRC_positivity_2020-03, PRC_positivity_2020-04, PRC_positivity_2020-05, PRC_positivity_2020-06, PRC_positivity_2020-07, PRC_positivity_2020-08, PRC_positivity_2020-09, PRC_positivity_2020-10, PRC_positivity_2020-11, PRC_positivity_2020-12, PRC_positivity_2021-01, PRC_positivity_2021-02, PRC_positivity_2021-03, PRC_positivity_2021-04, PRC_positivity_2021-05, PRC_positivity_2021-06, PRC_positivity_2021-07, PRC_positivity_2021-08, PRC_positivity_2021-09, PRC_positivity_2021-10, Poblacion_x.2, test_coverage_2020-03, test_coverage_2020-04, test_coverage_2020-05, test_coverage_2020-06, test_coverage_2020-07, test_coverage_2020-08, test_coverage_2020-09, test_coverage_2020-10, test_coverage_2020-11, test_coverage_2020-12, test_coverage_2021-01, test_coverage_2021-02, ...]
Index: []

[0 rows x 129 columns]

In [ ]:
df_2 = df_master.replace(0,np.nan).dropna(axis=1,how="all")
df_3 = df_2.replace(np.nan,0).dropna(axis=1,how="all")
df_master = df_3
df_master.drop(columns=['Poblacion_y.1'])
df_master

Region            Comuna  ...  BAC_2021-06  BAC_2021-07
0    Arica y Parinacota             Arica  ...       86.425       87.700
1    Arica y Parinacota         Camarones  ...       95.950       96.875
2    Arica y Parinacota     General Lagos  ...       96.075       97.050
3    Arica y Parinacota             Putre  ...       90.325       87.525
4              Tarapacá     Alto Hospicio  ...       84.025       86.075
..                  ...               ...  ...          ...          ...
339          Magallanes      Punta Arenas  ...       81.725       81.650
340          Magallanes         Rio Verde  ...       25.000       79.175
341          Magallanes      San Gregorio  ...       41.675       70.825
342          Magallanes          Timaukel  ...       75.000       44.225
343          Magallanes  Torres del Paine  ...       95.825       90.000

[344 rows x 92 columns]

Hacemos una descripción de los datos
 a usar

In [ ]:
df_master.describe()

Codigo comuna    Poblacion_x  ...  BAC_2021-06  BAC_2021-07
count     344.000000     344.000000  ...   344.000000   344.000000
mean     9019.200581   56562.534884  ...    66.104506    68.369985
std      3823.451166   89102.350327  ...    14.193924    14.482872
min      1101.000000     211.000000  ...    16.450000     8.325000
25%      6108.750000   10013.750000  ...    57.668750    61.012500
50%      8312.500000   20147.000000  ...    66.662500    69.812500
75%     13103.250000   56106.500000  ...    75.987500    78.512500
max     16305.000000  645909.000000  ...    98.375000    97.050000

[8 rows x 90 columns]

Dividimos las columas correspondientes por población

In [ ]:
for i, row in enumerate(df_master.iterrows()):
  for p, column in enumerate(df_master):
    if 'cases' in column or 'deceased' in column:
      cant_poblacion = df_master.loc[[i],['Poblacion_x']]
      df_master.at[i, column]= df_master.iloc[[i],[p]].squeeze() / cant_poblacion.squeeze()  
  

In [ ]:
df_master.describe()

Codigo comuna    Poblacion_x  ...  BAC_2021-06  BAC_2021-07
count     344.000000     344.000000  ...   344.000000   344.000000
mean     9019.200581   56562.534884  ...    66.104506    68.369985
std      3823.451166   89102.350327  ...    14.193924    14.482872
min      1101.000000     211.000000  ...    16.450000     8.325000
25%      6108.750000   10013.750000  ...    57.668750    61.012500
50%      8312.500000   20147.000000  ...    66.662500    69.812500
75%     13103.250000   56106.500000  ...    75.987500    78.512500
max     16305.000000  645909.000000  ...    98.375000    97.050000

[8 rows x 90 columns]

El dataframe final cuenta con 90 columnas y 344 filas.

Para ver un poco la distribución de los datos, vamos a mostrar la evolución de los casos totales.

Obtenemos el dataframe para hacer el análisis de la región metropolitana. 

In [ ]:
df_master_metropolitana = df_master[df_master['Region'] == 'Metropolitana']
df_master_metropolitana

Region               Comuna  ...  BAC_2021-06  BAC_2021-07
82   Metropolitana                Alhue  ...       76.725       79.425
83   Metropolitana                 Buin  ...       61.600       61.675
84   Metropolitana      Calera de Tango  ...       55.200       52.175
85   Metropolitana            Cerrillos  ...       71.450       71.325
86   Metropolitana          Cerro Navia  ...       63.875       64.950
87   Metropolitana               Colina  ...       68.100       76.500
88   Metropolitana             Conchali  ...       42.000       44.975
89   Metropolitana             Curacavi  ...       62.850       65.525
90   Metropolitana            El Bosque  ...       49.275       48.900
91   Metropolitana             El Monte  ...       65.125       67.400
92   Metropolitana     Estacion Central  ...       68.650       72.700
93   Metropolitana           Huechuraba  ...       58.050       58.625
94   Metropolitana        Independencia  ...       58.200       60.900
95   Metropolitana        Isla de Maipo  ...       72.250       74.675
96   Metropolitana          La Cisterna  ...       61.325       62.700
97   Metropolitana           La Florida  ...       45.550       50.675
98   Metropolitana            La Granja  ...       31.325       35.350
99   Metropolitana           La Pintana  ...       44.275       45.450
100  Metropolitana             La Reina  ...       66.300       66.900
101  Metropolitana                Lampa  ...       44.000       48.825
102  Metropolitana           Las Condes  ...       62.750       69.300
103  Metropolitana         Lo Barnechea  ...       77.475       79.900
104  Metropolitana            Lo Espejo  ...       49.300       56.450
105  Metropolitana             Lo Prado  ...       54.600       56.700
106  Metropolitana                Macul  ...       60.000       62.900
107  Metropolitana                Maipu  ...       67.050       72.625
108  Metropolitana          Maria Pinto  ...       74.875       78.200
109  Metropolitana            Melipilla  ...       56.375       61.575
110  Metropolitana                Nunoa  ...       67.300       70.500
111  Metropolitana        Padre Hurtado  ...       58.950       65.650
112  Metropolitana                Paine  ...       65.700       70.150
113  Metropolitana  Pedro Aguirre Cerda  ...       49.450       48.850
114  Metropolitana             Penaflor  ...       67.650       71.800
115  Metropolitana            Penalolen  ...       46.550       50.100
116  Metropolitana               Pirque  ...       52.675       61.300
117  Metropolitana          Providencia  ...       71.500       74.325
118  Metropolitana             Pudahuel  ...       50.875       55.875
119  Metropolitana          Puente Alto  ...       47.925       51.125
120  Metropolitana            Quilicura  ...       46.550       49.400
121  Metropolitana        Quinta Normal  ...       59.100       61.800
122  Metropolitana             Recoleta  ...       46.600       49.175
123  Metropolitana                Renca  ...       43.700       50.850
124  Metropolitana         San Bernardo  ...       53.550       56.825
125  Metropolitana          San Joaquin  ...       53.350       61.050
126  Metropolitana    San Jose de Maipo  ...       53.725       59.625
127  Metropolitana           San Miguel  ...       60.450       65.200
128  Metropolitana            San Pedro  ...       27.925       35.375
129  Metropolitana            San Ramon  ...       33.325       39.775
130  Metropolitana             Santiago  ...       68.050       71.850
131  Metropolitana            Talagante  ...       66.800       71.625
132  Metropolitana               Tiltil  ...       40.200       48.225
133  Metropolitana             Vitacura  ...       77.500       80.150

[52 rows x 92 columns]

In [ ]:
df_master_metropolitana.describe()

Codigo comuna    Poblacion_x  ...  BAC_2021-06  BAC_2021-07
count      52.000000      52.000000  ...    52.000000    52.000000
mean    13238.076923  156251.384615  ...    57.268269    60.921154
std       178.051309  135806.806340  ...    11.968484    11.571947
min     13101.000000    7405.000000  ...    27.925000    35.350000
25%     13113.750000   85574.000000  ...    48.937500    50.806250
50%     13126.500000  123316.500000  ...    58.575000    61.625000
75%     13401.250000  182598.500000  ...    66.862500    70.706250
max     13605.000000  645909.000000  ...    77.500000    80.150000

[8 rows x 90 columns]

Este dataframe cuenta también con 90 columnas, pero solo 52 filas que corresponden a las comunas de la región metropolitana. 

Como para la siguiente sección queremos reducir dimensionalidad y luego clusterizar, queremos encontrar un valor optimo de clusters de KMeans. Para eso definimos la siguiente función que elige según el SilhouetteScore.

In [ ]:
from sklearn.metrics import silhouette_score

def best_silhouette_score_clusters(X, k=30):
    j = 0
    max_score = -1
    for i in range(2, k + 1):
        kmeans = KMeans(n_clusters=i)
        pca_clusters = kmeans.fit_predict(X)
        score = silhouette_score(X, pca_clusters)
        if score > max_score:
            max_score = score
            j = i

    return (j, max_score)

def top_silhouette_score_clusters(X, k=30):
    clusters = []
    for i in range(2, k + 1):
        kmeans = KMeans(n_clusters=i)
        pca_clusters = kmeans.fit_predict(X)
        score = silhouette_score(X, pca_clusters)
        clusters.append((i, score))

    return sorted(clusters, key=lambda x: -x[1])[:4]

## 4. Análisis a nivel país

Definimos el X que usaremos de aquí en adelante. Nos quedamos solo con los valores numéricos del df

In [ ]:
X, y = df_master.drop(columns=['Comuna', 'Region']), df_master[['Region']]
X

Codigo comuna  Poblacion_x  ...  BAC_2021-06  BAC_2021-07
0          15101.0     247552.0  ...       86.425       87.700
1          15102.0       1233.0  ...       95.950       96.875
2          15202.0        810.0  ...       96.075       97.050
3          15201.0       2515.0  ...       90.325       87.525
4           1107.0     129999.0  ...       84.025       86.075
..             ...          ...  ...          ...          ...
339        12101.0     141984.0  ...       81.725       81.650
340        12103.0        211.0  ...       25.000       79.175
341        12104.0        681.0  ...       41.675       70.825
342        12303.0        282.0  ...       75.000       44.225
343        12402.0       1021.0  ...       95.825       90.000

[344 rows x 90 columns]

### 4.1 Usando PCA
Primero vamos a hacer reducción de dimensionalidad a dos dimensiones con PCA para graficar

In [ ]:
# Estandarizamos
std_scaler = StandardScaler()
X_std = std_scaler.fit_transform(X)

# Queremos reducir a las dimensiones a 2 
pca_2 = PCA(n_components=2)

# Guardamos el dataset pero ahora en su versión reducida en 2 dimensiones
X_PCA = pca_2.fit_transform(X_std)
pca_2.explained_variance_ratio_

array([0.26347949, 0.14232497])

Probamos clusterizando con **kmeans**

In [ ]:
best_silhouette_score_clusters(X_PCA)

(2, 0.5092307149511924)

A continuación un listado de los 4 mejores silhouette score para dos dimensiones

In [ ]:
top_silhouette_score_clusters(X_PCA)

[(2, 0.5092307149511924),
 (4, 0.3937331960139819),
 (29, 0.3768170872758968),
 (3, 0.37622815764458145)]

In [ ]:
# el mejor score es con 2 clusters para kmeans

kmeans2 = KMeans(n_clusters=2)
pca_clusters2 = kmeans2.fit_predict(X_PCA)
pca_clusters2

array([0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

Graficamos

In [ ]:
df_pca_kmeans = pd.concat([pd.DataFrame(X_PCA), pd.DataFrame(pca_clusters2), pd.DataFrame(y), df_master['Comuna']], axis=1)
df_pca_kmeans.set_axis(['dim1', 'dim2', 'cluster', 'region', 'comuna'], axis=1, inplace=True)
df_pca_kmeans

dim1      dim2  cluster              region            comuna
0     3.791244  1.045015        0  Arica y Parinacota             Arica
1    -2.488876  7.377346        1  Arica y Parinacota         Camarones
2     3.461578 -1.568894        1  Arica y Parinacota     General Lagos
3    -5.888086  3.277928        1  Arica y Parinacota             Putre
4     1.597426 -0.009610        1            Tarapacá     Alto Hospicio
..         ...       ...      ...                 ...               ...
339  10.273062 -6.972609        0          Magallanes      Punta Arenas
340  -5.685371 -1.163658        1          Magallanes         Rio Verde
341  -0.374341 -8.813634        1          Magallanes      San Gregorio
342  -2.982572 -6.918509        1          Magallanes          Timaukel
343 -10.133149  2.636717        1          Magallanes  Torres del Paine

[344 rows x 5 columns]

In [ ]:
px.scatter(df_pca_kmeans, x='dim1', y='dim2', hover_name='comuna', hover_data=['region'], color='cluster')

A continuación utilizamos PCA manteniendo el 95% de varianza

In [ ]:
# Estandarizamos
std_scaler = StandardScaler()
X_std = std_scaler.fit_transform(X)

# Queremos reducir manteniendo el 95%
pca = PCA(n_components=0.95)

# Guardamos el dataset pero ahora en su versión reducida
X_95 = pca.fit_transform(X_std)

print(X_95.shape)
pca.explained_variance_ratio_


(344, 26)


array([0.26347949, 0.14232497, 0.10887649, 0.08733959, 0.06175616,
       0.05044536, 0.02906438, 0.0242371 , 0.02267624, 0.02054941,
       0.01843255, 0.0151571 , 0.01341475, 0.01125614, 0.01072001,
       0.00907667, 0.00821724, 0.00810274, 0.00773285, 0.00674831,
       0.00637339, 0.00564579, 0.00533025, 0.00483152, 0.00456709,
       0.00420306])

Nos quedamos con 26 dimensiones para obtener el 0.95 de varianza 

In [ ]:
best_silhouette_score_clusters(X_95)

(2, 0.30759808190623933)

In [ ]:
top_silhouette_score_clusters(X_95)

[(2, 0.30434703046986034),
 (3, 0.18634479601206275),
 (7, 0.15691386930129497),
 (6, 0.15585546022116004)]

In [ ]:
# nos quedamos con 2 clusters que es el mejor score
kmeans95_2 = KMeans(n_clusters=2)
pca_clusters95 = kmeans95_2.fit_predict(X_95)
pca_clusters95

array([1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [ ]:
df_pca_kmeans = pd.concat([pd.DataFrame(X_95), pd.DataFrame(pca_clusters95), pd.DataFrame(y), df_master['Comuna']], axis=1)
df_pca_kmeans.set_axis(["dim" + str(i + 1) for i in range(X_95.shape[1])] + ["cluster", "region", "comuna"], axis=1, inplace=True)
df_pca_kmeans

dim1      dim2  ...              region            comuna
0     3.791244  1.045015  ...  Arica y Parinacota             Arica
1    -2.488876  7.377346  ...  Arica y Parinacota         Camarones
2     3.461578 -1.568894  ...  Arica y Parinacota     General Lagos
3    -5.888086  3.277928  ...  Arica y Parinacota             Putre
4     1.597426 -0.009610  ...            Tarapacá     Alto Hospicio
..         ...       ...  ...                 ...               ...
339  10.273062 -6.972609  ...          Magallanes      Punta Arenas
340  -5.685371 -1.163658  ...          Magallanes         Rio Verde
341  -0.374341 -8.813634  ...          Magallanes      San Gregorio
342  -2.982572 -6.918509  ...          Magallanes          Timaukel
343 -10.133149  2.636717  ...          Magallanes  Torres del Paine

[344 rows x 29 columns]

## 5. Análisis Región Metropolitana

In [ ]:
X_metrop, y_metrop = df_master_metropolitana.drop(columns=['Comuna', 'Region']), df_master_metropolitana[['Comuna']]

### 5.1 Usando PCA

In [ ]:
# Estandarizamos
std_scaler1 = StandardScaler()
X_metrop_std = std_scaler1.fit_transform(X_metrop)

# Queremos reducir las dimensiones a 2 para graficar algo
pca_2 = PCA(n_components=2)

# Guardamos el dataset pero ahora en su versión reducida en 2 dimensiones
X_metrop_PCA = pca_2.fit_transform(X_metrop_std)
pca_2.explained_variance_ratio_

array([0.38969806, 0.16337092])

Probamos clusterizando con **kmeans**

In [ ]:
best_silhouette_score_clusters(X_metrop_PCA)

(3, 0.4747927951565718)

In [ ]:
top_silhouette_score_clusters(X_metrop_PCA)

[(3, 0.4747927951565718),
 (16, 0.46074067717936246),
 (2, 0.4602819145316686),
 (18, 0.4492940638390651)]

In [ ]:
# Elegimos 3 clusters porque es el que da mejor score

kmeans_3 = KMeans(n_clusters=3)
pca_clusters_metrop = kmeans_3.fit_predict(X_metrop_PCA)
pca_clusters_metrop

array([0, 0, 0, 0, 1, 2, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 2, 0, 2, 2,
       1, 1, 0, 0, 0, 0, 2, 0, 0, 1, 0, 1, 0, 2, 1, 1, 0, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 2, 0, 0, 2], dtype=int32)

Graficamos

In [ ]:
df_pca_kmeans_metrop = pd.concat([pd.DataFrame(X_metrop_PCA), pd.DataFrame(pca_clusters_metrop), y_metrop.reset_index()['Comuna']], axis=1)
df_pca_kmeans_metrop.set_axis(['dim1', 'dim2', 'cluster', 'comuna'], axis=1, inplace=True)

In [ ]:
px.scatter(df_pca_kmeans_metrop, x='dim1', y='dim2', hover_name='comuna', color='cluster')

Ahora calculamos PCA con 0.95 

In [ ]:
# Estandarizamos
std_scaler = StandardScaler()
X_std = std_scaler.fit_transform(X_metrop)

# Queremos reducir manteniendo el 95%
pca_95 = PCA(n_components=0.95)

# Guardamos el dataset pero ahora en su versión reducida
X_metrop_PCA95 = pca_95.fit_transform(X_metrop_std)


print(X_metrop_PCA95.shape)
pca_95.explained_variance_ratio_

(52, 15)


array([0.38969806, 0.16337092, 0.11085838, 0.06883086, 0.04798883,
       0.04437071, 0.03110676, 0.02020835, 0.01546496, 0.01472199,
       0.01081111, 0.01001903, 0.00860689, 0.00780103, 0.00625493])

Para obtener el 0.95 de varianza nos quedamos con 15 dimensiones

In [ ]:
best_silhouette_score_clusters(X_metrop_PCA95)

(3, 0.2513083807118407)

In [ ]:
top_silhouette_score_clusters(X_metrop_PCA95)

[(2, 0.2509995974297051),
 (3, 0.2434473813324963),
 (4, 0.20401941788951025),
 (6, 0.17804420706984858)]

In [ ]:
# usamos 4 clusters porque tiene un buen score y coincide con la implementación del plan paso a paso del gobierno
kmeans_4 = KMeans(n_clusters=4)
pca_clusters_metrop = kmeans_4.fit_predict(X_metrop_PCA95)
pca_clusters_metrop

array([0, 0, 0, 0, 1, 3, 1, 0, 1, 0, 0, 3, 1, 0, 1, 3, 1, 1, 2, 3, 2, 2,
       1, 1, 0, 3, 0, 0, 2, 0, 0, 1, 0, 1, 0, 2, 3, 3, 3, 1, 1, 1, 3, 1,
       1, 1, 0, 1, 3, 0, 3, 2], dtype=int32)

In [ ]:
df_pca_kmeans_metrop = pd.concat([pd.DataFrame(X_metrop_PCA95), pd.DataFrame(pca_clusters_metrop), y_metrop.reset_index()['Comuna']], axis=1)
df_pca_kmeans_metrop.set_axis(["dim" + str(i + 1) for i in range(X_metrop_PCA95.shape[1])] + ["cluster", "comuna"], axis=1, inplace=True)

## 6. Visualización en Mapas

### 6.1 Región Metropolitana

In [ ]:
# Agregamos la columna de codigo de comuna al resultado de aplicar PCA y Kmeans a los datos
df_pca_kmeans_metrop['CodigoComuna'] = df_master_metropolitana['Codigo comuna'].reset_index(drop=True).astype(int)
# Y pasamos los clusters obtenidos a strings para que plotly los considere como categorias discretas
df_pca_kmeans_metrop['cluster'] = df_pca_kmeans_metrop['cluster'].astype(str)

In [ ]:
rm_geojson = 'https://raw.githubusercontent.com/caracena/chile-geojson/master/13.geojson'
fig = px.choropleth_mapbox(df_pca_kmeans_metrop, geojson=rm_geojson, locations='CodigoComuna', color='cluster',
                           hover_data=["comuna"],
                           mapbox_style="carto-positron",
                           featureidkey="properties.cod_comuna",
                           zoom=7, center = {"lat": -33.4489, "lon": -70.6693})

fig.show()

### 6.2 Todas las Comunas de Chile

In [ ]:
# Agregamos la columna de codigo de comuna al resultado de aplicar PCA y Kmeans a los datos
df_pca_kmeans['CodigoComuna'] = df_master['Codigo comuna'].reset_index(drop=True).astype(int)
# Y pasamos los clusters obtenidos a strings para que plotly los considere como categorias discretas
df_pca_kmeans['cluster'] = df_pca_kmeans['cluster'].astype(str)
df_pca_kmeans

dim1      dim2  ...            comuna  CodigoComuna
0     3.791244  1.045015  ...             Arica         15101
1    -2.488876  7.377346  ...         Camarones         15102
2     3.461578 -1.568894  ...     General Lagos         15202
3    -5.888086  3.277928  ...             Putre         15201
4     1.597426 -0.009610  ...     Alto Hospicio          1107
..         ...       ...  ...               ...           ...
339  10.273062 -6.972609  ...      Punta Arenas         12101
340  -5.685371 -1.163658  ...         Rio Verde         12103
341  -0.374341 -8.813634  ...      San Gregorio         12104
342  -2.982572 -6.918509  ...          Timaukel         12303
343 -10.133149  2.636717  ...  Torres del Paine         12402

[344 rows x 30 columns]

In [ ]:
chile_geojson = 'https://gist.githubusercontent.com/MarceloSaldias/8123ad8d4a78737123703159850ef8d0/raw/8cf696b51d1176aa1f5a4539eedfaf8a8c184259/All_Comunas.geojson'
fig = px.choropleth_mapbox(df_pca_kmeans, geojson=chile_geojson, locations='CodigoComuna', color='cluster',
                           hover_data=["comuna", "region"],
                           mapbox_style="carto-positron",
                           featureidkey="properties.cod_comuna",
                           zoom=3.5, center = {"lat": -40, "lon": -70.6693})

fig.update_layout(
    autosize=False,
    width=700,
    height=1000,
    margin=dict(
        l=50,
        r=50,
        b=20,
        t=20,
        pad=1
    ),
)

fig.show()

## 7. Conclusiones

A partir de los resultados, podemos observar que el análisis pareciera tener coincidencia con lo que esperábamos. Cuando realizamos el análisis con todas las comunas vemos que las de regiones (no metropolitana) quedan graficadas muy cerca y en el mismo cluster o en clusters muy cercanos. Con respecto a las comunas de la región metropolitana, los clusters coinciden con respecto a su nivel socieconómico, cantidad de población y localidad, por ejemplo comunas como Providencia, La Reina, Lo Barnechea, Las Condes, Vitacura y Ñuñoa quedan en el mismo cluster de `Kmeans`, mientras que comunas rurales como Curacaví, Isla de Maipo, Paine, Talagante y Pirque, entre otras, están en otro cluster. 


### Que se puede hacer con este modelo?  

Primero, puede servir para estudiar implementaciones de medidas ya sean preventivas o reactivas en ciertas comunas representativas de los clusters, analizar como funcionan y en base a eso decidir si implementar en el resto de las comunas del cluster.

Además, si bien actualmente, la mayoría de las comunas se encuentra en fase 4, eventualmente con las nuevas variantes o con otras enfermedades como la influenza,esta agrupación de comunas podría servir para tomar mejores medidas preventivas. Cabe destacar que en este modelo está entrenado con datos de COVID-19, por lo que en caso de utilizarlo para otras enfermedades o variantes que puedan tener otro comportamiento, esto es una buena aproximación inicial ,pero con el tiempo se debería ir calibrando con los datos respectivos.

### Posibles aplicaciones

Teniendo los datos respectivos,  cualquier  caso de uso que tenga relación con comportamiento de comunas, por ejemplo votaciones, puntajes PTU, entre otros.


